### PR TITLE
Automate triggering e2e workflow

### PR DESCRIPTION
**Pull request by russellb**
9842fde e2e: Auto-launch workflow via the `e2e-trigger` label
bbddb48 e2e: Add a comment to PRs with workflow link
7482de8 Makefile: Add `action-lint` target
044a0ff Makefile: Add spellcheck target
247faf9 docs: Add more explanation of when to run this job

commit 9842fdeae7da73d48283d439735daac1e5b269a3
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun May 19 04:56:39 2024 -0400

    e2e: Auto-launch workflow via the `e2e-trigger` label
    
    This change automates launching the `e2e` workflow against a PR in
    response to adding the `e2e-trigger` label. After launching the
    workflow, the label will be removed.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit bbddb486847b1817f9cb64779c019b0163d1e6a3
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun May 19 05:02:00 2024 -0400

    e2e: Add a comment to PRs with workflow link
    
    When the `e2e` workflow is run against a PR, add a comment to the PR
    automatically with a link to the workflow run details.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 7482de8af0d3b9ce4d6d9a69eb4784a49f085aca
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun May 19 05:06:00 2024 -0400

    Makefile: Add `action-lint` target
    
    CI runs `actionlint` on github workflow configuration. Add a
    `Makefile` target to run the same linting locally for easier local
    checking when making changes.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 044a0ff075eb4112b8efbd013d8ac8c401352094
Author: Russell Bryant <rbryant@redhat.com>
Date:   Thu May 30 16:27:46 2024 -0400

    Makefile: Add spellcheck target
    
    Make it easy to reproduce what is done in the spellcheck CI job.
    Also include some missing words highlighted by CI and reproduced using
    this target.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 247faf911089bcfd974427b567ce318deb0f1ddd
Author: Russell Bryant <rbryant@redhat.com>
Date:   Fri May 31 14:49:24 2024 -0400

    docs: Add more explanation of when to run this job
    
    This job only runs when manually triggered, so add a bit of guidance
    for when it is appropriate to use. This also explains why it is
    manually triggered only (cost).
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -31,6 +31,22 @@ jobs:
             echo "is_pr=false" >> "$GITHUB_OUTPUT"
           fi
 
+      - name: Add comment to PR
+        if: steps.check_pr.outputs.is_pr == 'true'
+        uses: actions/github-script@v5
```
Comment (nathan-weinberg) [inline .github/workflows/e2e.yml:36]:
```suggestion
        uses: actions/github-script@v5
```
We should probably harden this with a specific SHA

File: .spellcheck-en-custom.txt [hunk 2]
```diff
@@ -1,124 +1,125 @@
...
+submodules
+tatsu
+th
+tl
+tox
+triagers
+unquantized
+venv
+wikisql
+xcode
```
Comment (nathan-weinberg) [inline .spellcheck-en-custom.txt:125]: Spellcheck sort I think needs to be run?

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]

File: .spellcheck-en-custom.txt (hunk 2 above)
Comment (russellb) [reply .spellcheck-en-custom.txt:125]: funny enough, these changes are the result of running `make spellcheck-sort`. I only added one word.

File: .github/workflows/e2e.yml (hunk 1 above)
Comment (russellb) [reply .github/workflows/e2e.yml:36]: Is that really necessary for official actions from github? This is in line with how github documents usage. If github's own actions are compromised, that'd be pretty messed up.

@bjhargrave curious on your input here

Comment (bjhargrave) [reply .github/workflows/e2e.yml:36]: Best practice for security it to use SHAs (and dependabot to keep things up-to-date). SHAs also important for reproducibility. v5 will move so if you make a branch from here in the future using a SHA will use the exact same revision of the action as is used now. So even if you fully trust GitHub, SHAs are still a best practice.

Comment (russellb) [reply .github/workflows/e2e.yml]: Thanks to you both for the input.

I updated this to use a hash, instead.

**Review by nathan-weinberg — APPROVED** [review]
LGTM @russellb - do we have any way of gating who can run the job since it does cost us some money?

**Comment by russellb** [comment]
> LGTM @russellb - do we have any way of gating who can run the job since it does cost us some money?

Only people with write access can trigger it manually. Triagers can set labels so that opens it up a bit more. Otherwise people can’t run it. When the label is set it will only run once because the label gets removed too. That’s the theory anyway … I have to actually test this now that it’s merged!

**Review by leseb — COMMENTED** [review]
What should be the trigger to run the e2e job? I'm wondering if it should run before merging, let's say we have two approvals on a PR and CI is green, we then enter the final stage before merging, how about we run the e2e workflow during that stage? If the e2e job succeeds then we let Mergify merge. This way we still minimize the cost by helpfully running the e2e test once per PR. Also, it's something maintainers don't have to think about for each PR on whether it makes sense or not to run the e2e test and so nobody forgets to run it.

Another question, what is the current schedule on which this e2e job runs?

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -31,6 +31,22 @@ jobs:
...
+        if: steps.check_pr.outputs.is_pr == 'true'
+        uses: actions/github-script@60a0d83 # v7.0.1
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const issue_number = ${{ github.event.inputs.pr_or_branch }};
+            const repository = context.repo;
+            const url = `https://github.com/${repository.owner}/${repository.repo}/actions/runs/${process.env.GITHUB_RUN_ID}`;
+            const message = `e2e workflow launched on this PR: [View run](${url})`;
+            github.rest.issues.createComment({
```
Comment (bjhargrave) [inline .github/workflows/e2e.yml:44]: Why not just use the `gh` command available on the runner? 

`gh pr comment <prnumber> --body "<message>"`

Then you don't need to use actions/github-script. It can all be done in bash.

**Comment by russellb** [comment]
> What should be the trigger to run the e2e job? I'm wondering if it should run before merging, let's say we have two approvals on a PR and CI is green, we then enter the final stage before merging, how about we run the e2e workflow during that stage? If the e2e job succeeds then we let Mergify merge. This way we still minimize the cost by helpfully running the e2e test once per PR. Also, it's something maintainers don't have to think about for each PR on whether it makes sense or not to run the e2e test and so nobody forgets to run it.
> 
> Another question, what is the current schedule on which this e2e job runs?

It only runs when manually triggered. This PR was adding a second mechanism for that manual triggering that is accessible to a broader group. There's nowhere it runs automatically right now, mostly because it costs about $2 USD each time we run it, so I wanted to start slow. Ideally it'd run on any change that could break the e2e flow, but I wanted to do some math on it before making it automatic so we had some budget expectations.

**Comment by leseb** [comment]
> > What should be the trigger to run the e2e job? I'm wondering if it should run before merging, let's say we have two approvals on a PR and CI is green, we then enter the final stage before merging, how about we run the e2e workflow during that stage? If the e2e job succeeds then we let Mergify merge. This way we still minimize the cost by helpfully running the e2e test once per PR. Also, it's something maintainers don't have to think about for each PR on whether it makes sense or not to run the e2e test and so nobody forgets to run it.
> > Another question, what is the current schedule on which this e2e job runs?
> 
> It only runs when manually triggered. This PR was adding a second mechanism for that manual triggering that is accessible to a broader group. There's nowhere it runs automatically right now, mostly because it costs about $2 USD each time we run it, so I wanted to start slow. Ideally it'd run on any change that could break the e2e flow, but I wanted to do some math on it before making it automatic so we had some budget expectations.

Understood. I don't know what the budget is for this CI job. However, running it right before merging any PR that has `*.py` touched seems reasonable to me, the best case is $2 per PR... Having this in place allows great coverage. If the budget allows it, I'd like to push for running this before each merge. Thanks!

**Review by leseb — CHANGES_REQUESTED** [review]
Holding off until requested changes are addressed. Thanks!

**Comment by russellb** [comment]
> > > What should be the trigger to run the e2e job? I'm wondering if it should run before merging, let's say we have two approvals on a PR and CI is green, we then enter the final stage before merging, how about we run the e2e workflow during that stage? If the e2e job succeeds then we let Mergify merge. This way we still minimize the cost by helpfully running the e2e test once per PR. Also, it's something maintainers don't have to think about for each PR on whether it makes sense or not to run the e2e test and so nobody forgets to run it.
> > > Another question, what is the current schedule on which this e2e job runs?
> > 
> > 
> > It only runs when manually triggered. This PR was adding a second mechanism for that manual triggering that is accessible to a broader group. There's nowhere it runs automatically right now, mostly because it costs about $2 USD each time we run it, so I wanted to start slow. Ideally it'd run on any change that could break the e2e flow, but I wanted to do some math on it before making it automatic so we had some budget expectations.
> 
> Understood. I don't know what the budget is for this CI job. However, running it right before merging any PR that has `*.py` touched seems reasonable to me, the best case is $2 per PR... Having this in place allows great coverage. If the budget allows it, I'd like to push for running this before each merge. Thanks!

I don't intend to change that in this PR -- only to allow manual triggering from a slightly wider audience.

File: Makefile [hunk 4]
```diff
@@ -188,3 +188,18 @@ check-engine:
 .PHONY: check-rocm
 check-rocm:
 	@command -v rocm &> /dev/null || (echo "'rocm' is not installed" && exit 1)
+
+.PHONY: action-lint
+action-lint: ## Lint GitHub Action workflows
+	$(ECHO_PREFIX) printf "  %-12s .github/...\n" "[ACTION LINT]"
+	$(CMD_PREFIX) if ! which actionlint $(PIPE_DEV_NULL) ; then \
+		echo "Please install actionlint." ; \
```
[the remark this answers is not in the document]
Comment (russellb) [reply Makefile:196]: because there's no format string usage here

File: .github/workflows/e2e.yml [hunk 5]
```diff
@@ -31,6 +31,22 @@ jobs:
             echo "is_pr=false" >> "$GITHUB_OUTPUT"
           fi
 
+      - name: Add comment to PR
+        if: steps.check_pr.outputs.is_pr == 'true'
+        uses: actions/github-script@60a0d83 # v7.0.1
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const issue_number = ${{ github.event.inputs.pr_or_branch }};
```
[the remark this answers is not in the document]
Comment (russellb) [reply .github/workflows/e2e.yml:40]: I'm going to redo this chunk to use the `gh` command so this line will go away

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (russellb) [reply .github/workflows/e2e.yml:44]: ok.

File: Makefile [hunk 6]
```diff
@@ -188,3 +188,18 @@ check-engine:
 .PHONY: check-rocm
 check-rocm:
 	@command -v rocm &> /dev/null || (echo "'rocm' is not installed" && exit 1)
+
+.PHONY: action-lint
+action-lint: ## Lint GitHub Action workflows
+	$(ECHO_PREFIX) printf "  %-12s .github/...\n" "[ACTION LINT]"
+	$(CMD_PREFIX) if ! which actionlint $(PIPE_DEV_NULL) ; then \
```
[the remark this answers is not in the document]
Comment (russellb) [reply Makefile:200]: both seem to work fine?

File: docs/ci.md [hunk 7]
```diff
@@ -2,13 +2,19 @@
 
 ## End-to-end CI Job
 
-This CI job is manually triggered by `instructlab` repo maintainers. It runs as
-much of the `ilab` workflow as it can on the GPU-enabled worker we have
-available through GitHub Actions.
+This CI job is manually triggered by `instructlab` repo triagers or maintainers.
```
[the remark this answers is not in the document]
Comment (russellb) [reply docs/ci.md:5]: whenever someone thinks a change could benefit from an end-to-end workflow test?  I'm not sure what to write that's super helpful.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: .spellcheck-en-custom.txt [hunk 8]
```diff
@@ -1,129 +1,137 @@
+=======
+>>>>>>> 91bdd6d (e2e: Auto-launch workflow via the `e2e-trigger` label)
```
[the remark this answers is not in the document]
Comment (russellb) [reply .spellcheck-en-custom.txt:2]: oops :)

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (russellb) [reply Makefile:200]: I understand - I was looking for an explanation of why.  I looked it up, though.

tl;dr - it's a POSIX standard.  `which` is not (though seems available everywhere in practice ...)

File: .github/workflows/e2e.yml [hunk 9]
```diff
@@ -31,6 +31,12 @@ jobs:
             echo "is_pr=false" >> "$GITHUB_OUTPUT"
           fi
 
+      - name: Add comment to PR
+        id: pr_comment
+        if: steps.check_pr.outputs.is_pr == 'true'
+        run: |
+          gh pr comment "${{ github.event.inputs.pr_or_branch }}" -b "e2e workflow launched on this PR: [View run](https://github.com/${repository.owner}/${repository.repo}/actions/runs/${process.env.GITHUB_RUN_ID}"
```
[the remark this answers is not in the document]
Comment (russellb) [reply .github/workflows/e2e.yml]: Thanks for the edit!  I applied it.

File: Makefile [hunk 10]
```diff
@@ -188,3 +192,19 @@ check-engine:
...
+	$(CMD_PREFIX) if ! command -v actionlint $(PIPE_DEV_NULL) ; then \
+		echo "Please install actionlint." ; \
+		echo "go install github.com/rhysd/actionlint/cmd/actionlint@latest" ; \
+		exit 1 ; \
+	fi
+	$(CMD_PREFIX) if ! command -v shellcheck $(PIPE_DEV_NULL) ; then \
+		echo "Please install shellcheck." ; \
+		echo "https://github.com/koalaman/shellcheck#user-content-installing" ; \
+		exit 1 ; \
+	fi
```
Comment (leseb) [inline Makefile:209]: Why do we need shellcheck here? I don't see it being called.

File: docs/ci.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (leseb) [reply docs/ci.md:5]: I guess my point is that not all maintainers might know when to trigger it so a little guidance will help.

File: Makefile (hunk 10 above)
Comment (bjhargrave) [reply Makefile:209]: actionlint will use shellcheck if installed to check the bash scripts in the workflow files. We want this.

**Review by bjhargrave — APPROVED** [review]

**Review by nathan-weinberg — APPROVED** [review]

File: .github/workflows/e2e.yml [hunk 11]
```diff
@@ -31,6 +31,12 @@ jobs:
             echo "is_pr=false" >> "$GITHUB_OUTPUT"
           fi
 
+      - name: Add comment to PR
+        id: pr_comment
+        if: steps.check_pr.outputs.is_pr == 'true'
+        run: |
+          gh pr comment "${{ github.event.inputs.pr_or_branch }}" -b "e2e workflow launched on this PR: [View run](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"
```
Comment (bjhargrave) [inline .github/workflows/e2e.yml:38]:
```suggestion
          gh pr comment "${{ github.event.inputs.pr_or_branch }}" -b "e2e workflow launched on this PR: [View run](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }})"
```

Oops. Forgot the close paren for the md link.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e.yml (hunk 11 above)
Comment (russellb) [reply .github/workflows/e2e.yml:38]: great catch!

I wish this was easier to test prior to merging ... i fully expect a fixup PR may come after :(

File: docs/ci.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (russellb) [reply docs/ci.md:5]: Yeah, fair enough -- let me give it some thought.  I'll come up with something!

File: .github/workflows/e2e.yml (hunk 11 above)
Comment (russellb) [reply .github/workflows/e2e.yml:38]: I more meant the Mergify part where the workflow gets triggered via the label. That's kind of confusing since I was commenting on the workflow side.

File: docs/ci.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (russellb) [reply docs/ci.md:5]: I added some text, let me know if you think it's helpful!

File: .github/workflows/e2e.yml (hunk 11 above)
Comment (russellb) [reply .github/workflows/e2e.yml]: fixed the missing paren!

File: .github/workflows/e2e.yml [hunk 12]
```diff
@@ -31,6 +31,12 @@ jobs:
             echo "is_pr=false" >> "$GITHUB_OUTPUT"
           fi
 
+      - name: Add comment to PR
+        id: pr_comment
+        if: steps.check_pr.outputs.is_pr == 'true'
+        run: |
+          gh pr comment "${{ github.event.inputs.pr_or_branch }}" -b "e2e workflow launched on this PR: [View run](https://github.com/${repository.owner}/${repository.repo}/actions/runs/${process.env.GITHUB_RUN_ID})"
```
Comment (bjhargrave) [inline .github/workflows/e2e.yml:38]:
```suggestion
          gh pr comment "${{ github.event.inputs.pr_or_branch }}" -b "e2e workflow launched on this PR: [View run](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }})"
```

This change got lost somehow.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e.yml (hunk 12 above)
Comment (russellb) [reply .github/workflows/e2e.yml:38]: sigh, sorry ... i broke it with a rebase and forced push I'm sure ...

**Review by bjhargrave — APPROVED** [review]

**Comment by russellb** [comment]
@bjhargrave thank you for your diligence in review here!

File: docs/ci.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (leseb) [reply docs/ci.md:5]: Thanks!

**Review by leseb — APPROVED** [review]
